pkg/proxy/service/cluster_service_test.go

```diff
@@ -107,318 +107,410 @@ func (s *TestClusterServiceSuite) TestGetClusterAccess() {
+						s.assertClusterAccess(access.NewClusterAccess(*expectedURL, expectedToken, "smith"), ca)
+						assert.Equal(s.T(), "smith2", ca.Username())
+					})
+
+					s.Run("cluster access correct when using workspace context", func() {
+						// when
+						ca, err := svc.GetClusterAccess("789-ready", "", "smith2", "", publicViewerEnabled) // workspace-context specified
+
+						// then
+						require.NoError(s.T(), err)
+						require.NotNil(s.T(), ca)
+						expectedURL, err := url.Parse("https://api.endpoint.member-2.com:6443")
+						require.NoError(s.T(), err)
+						s.assertClusterAccess(access.NewClusterAccess(*expectedURL, expectedToken, "smith"), ca)
+						assert.Equal(s.T(), "smith2", ca.Username())
+
+						s.Run("another workspace on another cluster", func() {
+							// when
+							ca, err := svc.GetClusterAccess("789-ready", "", "teamspace", "", publicViewerEnabled) // workspace-context specified
+
+							// then
+							require.NoError(s.T(), err)
+							require.NotNil(s.T(), ca)
+							expectedURL, err := url.Parse("https://api.endpoint.member-1.com:6443")
+							require.NoError(s.T(), err)
+							s.assertClusterAccess(access.NewClusterAccess(*expectedURL, "def456", "smith"), ca)
+							assert.Equal(s.T(), "smith2", ca.Username())
+						})
+					})
 				})
 			})
 		})
+	}
 
-		s.Run("verify cluster access no route", func() {
-			memberClient.MockGet = nil
-			expectedToken := "abc123" // should match member 2 bearer token
-
+	// public-viewer specific tests
+	s.Run("user is public-viewer", func() {
+		s.Run("has no default workspace", func() {
 			// when
-			ca, err := svc.GetClusterAccess("789-ready", "", "", "")
+			ca, err := svc.GetClusterAccess("", toolchainv1alpha1.KubesawAuthenticatedUsername, "", "", true)
 
 			// then
-			require.NoError(s.T(), err)
-			require.NotNil(s.T(), ca)
-			expectedURL, err := url.Parse("https://api.endpoint.member-2.com:6443")
-			require.NoError(s.T(), err)
-			assert.Equal(s.T(), "smith2", ca.Username())
+			require.EqualError(s.T(), err, "user is not provisioned (yet)")
+			require.Nil(s.T(), ca)
+		})
 
-			s.assertClusterAccess(access.NewClusterAccess(*expectedURL, expectedToken, ""), ca)
+		s.Run("get workspace by name", func() {
+			svc := service.NewMemberClusterService(
+				fake.MemberClusterServiceContext{
+					Client: s,
+					Svcs:   s.Application,
+				},
+				func(si *service.ServiceImpl) {
+					si.GetMembersFunc = func(_ ...commoncluster.Condition) []*commoncluster.CachedToolchainCluster {
+						return s.memberClusters()
+					}
+				},
+			)
 
-			s.Run("cluster access correct when username provided", func() {
+			s.Run("public-viewer is disabled", func() {
 				// when
-				ca, err := svc.GetClusterAccess("", "smith@", "", "")
+				ca, err := svc.GetClusterAccess("", toolchainv1alpha1.KubesawAuthenticatedUsername, "smith2", "", false)
 
 				// then
-				require.NoError(s.T(), err)
-				require.NotNil(s.T(), ca)
+				require.Error(s.T(), err)
```

Comment:
It's minor but using `require.EqualError()` is usually better IMO. So you can check that you get the exact error you are expecting and not something unrelated to the test case.